common/djangoapps/student/models.py

```diff
@@ -3116,9 +3120,144 @@ class AccountRecoveryConfiguration(ConfigurationModel):
+
+    The first_day_of_streak and last_day_of_streak fields are used to
+    control celebration of the streak feature.
+    A streak is when a learner visits the learning MFE N days in a row.
+    The business logic of streaks for a 3 day streak and 1 day break is the following:
+    1. Each streak should be celebrated exactly once, once the learner has completed the streak.
+    2. If a learner misses enough days to count as a break, the streak resets back to 0.
+    3. The streak is measured against the learner's configured timezone
+    4. We keep track of the total length of the streak, so there is a possibility in the future
+    to add multiple celebrations for longer streaks.
+    5. We keep track of the longest_ever_streak field for potential future use for badging purposes.
+
+    .. no_pii:
+    """
+    user = models.OneToOneField(User, models.CASCADE, related_name='celebration')
+    first_day_of_streak = models.DateField(default=None, null=True, blank=True)
+    last_day_of_streak = models.DateField(default=None, null=True, blank=True)
+    longest_ever_streak = models.IntegerField(default=0)
+
+    def __str__(self):
+        return (
+            "[UserCelebration] user: {}; first_day_of_streak {};"
+            " last_day_of_streak {}; longest_ever_streak {};"
+        ).format(self.user.username, self.first_day_of_streak,
+                 self.last_day_of_streak, self.longest_ever_streak)
+
+    @classmethod
+    def streak_lengths_to_celebrate(cls):
+        return [3]
+
+    @classmethod
+    def streak_break_length(cls):
+        return 1
+
+    @classmethod
+    def _get_now(self):
+        """ Retrieve the value for the current datetime in the user's timezone """
+        # importing here to avoid a circular import
+        from lms.djangoapps.courseware.context_processor import user_timezone_locale_prefs
+        user_timezone_locale = user_timezone_locale_prefs(crum.get_current_request())
+        user_timezone = timezone(user_timezone_locale['user_timezone'] or str(UTC))
+        return user_timezone.localize(datetime.now())
+
+    def _calculate_streak_updates(self):
+        """ Calculate the updates that should me applied to the streak fields of the provided celebration """
+        first_day_of_streak = self.first_day_of_streak
+        last_day_of_streak = self.last_day_of_streak
+        streak_lengths_to_celebrate = self.streak_lengths_to_celebrate()
+        streak_break_length = self.streak_break_length()
+
+        today = self.__class__._get_now().date()  # pylint: disable=protected-access
+        streak_length_to_celebrate = None
+
+        first_ever_streak = None is first_day_of_streak is last_day_of_streak
+        inside_active_streak = last_day_of_streak and first_day_of_streak
+        can_start_new_streak = last_day_of_streak and last_day_of_streak + timedelta(days=streak_break_length) < today
+        already_celebrated_today = last_day_of_streak == today
```

Comment:
Can we rename this variable to something that doesn't seem to indicate we celebrate on other days? Like reading this, it feels weird that it would be True on days 2 or 8, etc.

I'm blanking on what a better name might be, but I'll try and think through an option. Just seeing how it's used only when it's a day to be celebrated.

Reply:
Yeah, I get the intention of the variable. My point is just that the implication of `already celebrated today` makes it sound like we could be celebrating on any day, when really it means more of a `we already updated this today`. Sorry for it not being clear. Maybe another way of trying to explain it is that it feels weird to me that on day 2 of a streak, you could have a variable titled `already_celebrated_today` having the value of `True` even though there wasn't a celebration that day.

Maybe a switch to `already_updated = last_day_of_streak == today`. Then you might also be able to short circuit using this check where if we already updated, you can just return the appropriate values and can still use this variable for your check later on.

Reply:
ah ok i understand, I moved the variable further up in response to a different comment but it took on a bit of a different meaning in the new spot
updated the name